### ISSUE TITLE
contract_metric(metric) in tensor module contracts indices of any TensorIndexType

### ISSUE THREAD

**Issue by pashilkar** [sympy/sympy#26289]
```python

from sympy import symbols
from sympy.tensor.tensor import tensor_indices, TensorIndexType

#dimensions d
d = symbols('d', positive=True, integer=True)

# Define 2 TensorIndexTypes:
# 'Lorentz' with metric 'g' &
# 'Minkowski' with metric 'η'.
Lorentz = TensorIndexType('Lorentz', dummy_name='L', dim = d, metric_name='g')
g = Lorentz.metric
Minkowski = TensorIndexType('Minkowski', dummy_name='M', dim = d, metric_name='η')
η = Minkowski.metric

# Define 3 indices of type 'Lorentz'.
μ, ν, σ = tensor_indices("μ, ν, σ", Lorentz)

# Contracting with η gives same result as contracting with g!
(η(μ, σ) * g(-ν, -σ)).contract_metric(η) - (g(μ, σ) * g(-ν, -σ)).contract_metric(g)
# output >> 0

```

contract_metric(η) should contract only Minkowski indices. But when applied to $η^{μσ} g_{νσ}$ with Lorentz indices, it outputs ${g_ν}^μ$.
contract_metric(g), on the other hand, works as expected on $g^{μσ} g_{νσ}$ outputting ${g_ν}^μ$.

**Comment by AshutoshRajora** [comment]
so basically,
The contract_metric method should contract only indices associated with the same metric. But, it appears that it's also contracting Lorentz indices when applied to the Minkowski metric.
have I got you right?

**Comment by pashilkar** [comment]
Yes.
η[Minkowski, Minkowski] should contract only Minkowski tensor indices.

**Comment by Kishore96in** [comment]
Does it make sense to pass `Lorentz` indices to `η`?
```
In [15]: η.index_types
Out[15]: [Minkowski, Minkowski]
```
It seems to me that TensorHead should raise an error if passed an index which does not match the corresponding entry in `index_types`.

**Comment by pashilkar** [comment]
Yes, a Tensor can have indices of any type, `η[Lorentz, Lorentz], η[Minkowski, Minkowski]` and even `η[Lorentz, Minkowski]` (e.g. in Tetrad Formulation of Einstein's General Relativity).
The issue here is with `contract_metric()`.
When we have two or more metrics `g`, `η` etc. (e.g., in Bimetric Theories), only one metric can be used to contract, raise and lower indices. In my case that metric is `g`.
`contract_metric()` should work with only one metric.

I'm sorry for misleading you with my last reply which was wrong.
> Yes. η[Minkowski, Minkowski] should contract only Minkowski tensor indices.

**Comment by pashilkar** [comment]
This is a followup to my previous comment...

Consider the following use case:
Let `g` & `η` be two metrics in spacetime. If `g` is taken to be the physical metric, then tensor contraction and raising/lowering of indices is performed using `g`.

```
α, β, γ = tensor_indices("α, β, γ", Lorentz)
a, b, c = tensor_indices("a, b, c", Minkowski)

```

Following is the expected behaviour of contract_metric()...

```
In : (η(-α, -γ) * g(γ, β)).contract_metric()
Out: η(-α, β)

In : (η(-a, -c) * g(c, b)).contract_metric()
Out: η(-a, b)

In : (η(-a, -γ) * g(γ, b)).contract_metric()
Out: η(-a, b)

In : (η(-α, -γ) * g(γ, b)).contract_metric()
Out: η(-α, b)

In : (η(-α, -c) * g(c, b)).contract_metric()
Out: η(-α, b)

In : (η(-α, -γ) * η(γ, β)).contract_metric()
Out: η(-α, -γ) * η(γ, β)
```

**Comment by pashilkar** [comment]
This issue should be closed as it stems from my misunderstanding regarding the relationship between ``TensorIndexType`, `metric` and `TensorIndex`. 

There is nothing to fix in the current implementation.

**Comment by pashilkar** [comment]
This issue should be closed as it stems from my misunderstanding regarding the relationship between ``TensorIndexType`, `metric` and `TensorIndex`. 

There is nothing to fix in the current implementation.
